shared/middleware/helpers/sharded_redis_factory.rb

```diff
@@ -0,0 +1,63 @@
+require 'redis-slave-read'
+require 'consistent_hashing'
+
+# Helper for getting a correctly configured Redis client for a particular
+# shard key.  Can be configured for any number of shards, where each shard
+# can be a single Redis node or a replication group where reads will be
+# distributed to read-replicas.
+class ShardedRedisFactory
+  # @param [Hash<'master':String, 'read_replicas':String[]>[]] config
+  # The set of Redis node URLs to be used in the current environment, provided
+  # in the following format:
+  # [
+  #   {
+  #     'master': 'redis://master1',
+  #     'read_replicas': [
+  #       'redis://master1replica1',
+  #       'redis://master1replica2'
+  #     ]
+  #   },
+  #   {
+  #     'master': 'redis://master2',
+  #     'read_replicas': [
+  #       'redis://master2replica1',
+  #       'redis://master2replica2'
+  #     ]
+  #   }
+  # ]
+  #
+  # The read_replicas key is optional and you can specify any number of groups.
+  #
+  # @param [Proc] new_redis_proc (optional) is provided as a hook to allow
+  # us to construct fake Redis instances in test.
+  def initialize(shards, new_redis_proc = proc {|url| Redis.new(url: url)})
```

Comment:
note: you can also use the "stab operator" lambda syntax `new_redis_proc = ->(url) {Redis.new(url: url)}`

Either way is fine

Reply:
Since procs and lambdas are slightly different and I'm not super familiar with those differences, I'm going to leave this as-is for now.